### ISSUE TITLE
Make no distinction between arm64-darwin-21, arm64-darwin-22 and arm64-darwin-23

### ISSUE THREAD

**Issue by sandstrom** [rubygems/rubygems#7407]
When running macOS with an ARM processor, our `Gemfile.lock` looks like this:

```
PLATFORMS
  aarch64-linux
  arm64-darwin-21
  arm64-darwin-22
  arm64-darwin-23
  arm64-linux
  x86_64-darwin-21
  x86_64-darwin-22
  x86_64-linux
```

This doesn't make sense to me, there is no discernible difference between 21/22/23.

I think it would make more sense if it looked like this:

```
PLATFORMS
  aarch64-linux
  arm64-darwin
  arm64-linux
  x86_64-darwin
  x86_64-linux
```

*(i.e. that the mac platform wouldn't get a bump every year)*

### Semi-related

- https://github.com/rubygems/rubygems/issues/5338

**Comment by gstokkink** [comment]
Agree with this, not sure what the purpose is of also including the Darwin kernel version in the lockfile, as it's not done for the other architectures either?

**Comment by hsbt** [comment]
We can easily normalize that with https://github.com/rubygems/rubygems/blob/master/lib/rubygems/platform.rb#L118.

But I'm not sure what breaking impact with this change. Does anyone survey that?

**Comment by sandstrom** [comment]
I think unknown platforms are just ignored.

## Any platform allowed(?)

So for example this platform list in `Gemfile.lock`:

```
PLATFORMS
  aarch64-linux
  arm64-darwin-21
  arm64-darwin-22
  arm64-darwin-23
  anything-goes-whatever-works
  arm64-linux
  x86_64-darwin-21
  x86_64-darwin-22
  x86_64-linux
```

Would not cause any trouble when running `bundle install`. The only thing it does is sorting it alphabetically.

## Automatic cleanup

However, we would be left with some "junk platforms" in the gem lockfile.

Perhaps a temporary addition to `bundle install` that would look for the pattern `arm64-darwin-[/d]{2}` and replace them with `arm64-darmwin` to help with automatic cleanup. Such temporary addition could be removed after e.g. ~1 year.

**Comment by deivid-rodriguez** [comment]
> Agree with this, not sure what the purpose is of also including the Darwin kernel version in the lockfile, as it's not done for the other architectures either?

We just parse what Ruby gives as as `RbConfig::CONFIG["arch"]`.

I agree that Darwin versions are not really different in practice? But not sure. I know people have actually released platform specific gems for different darwin versions but maybe they are actually all equal in code, and that's just due to how RubyGems and Bundler handle platforms right now. I'm also not sure about backwards compatibility.

One thing to note is that since recently, if no darwin version specific precompiled gems are found among resolved gems, Bundler won't lock specific version platforms, but the generic one, like requested here. So this is only a limited problem right now because most popular gems releasing precompiled versions don't ship `-darwin-<version>` variants, but just generic darwin variants.

**Comment by starburst997** [comment]
Just wanted to chime in as I encountered [an issue](https://github.com/ruby/setup-ruby/pull/702) related to this.

I have a Github Action workflow that worked no problem that uses [ruby/setup-ruby](https://github.com/ruby/setup-ruby) with cache enabled and a `Gemfile.lock` that run on `macos-13` runners. Suddenly, without change, I encountered this error:

```
Your bundle only supports platforms ["x64-mingw-ucrt", "x86_64-linux", "x86_64-darwin-21"] but your
local platform is x86_64-darwin-22. Add the current platform to the lockfile
with
`bundle lock --add-platform x86_64-darwin-22` and try again.
```

Now, `macos-13` are "supposed" to be `22` anyway, the reason why `21` used to work is because that number is embedded at compile time based on whichever platform was used to compile it. `setup-ruby` [previously used](https://github.com/ruby/setup-ruby/actions/runs/13049552263/job/36406659966#step:4:7) `macos-12` runner to build, but [now are using](https://github.com/ruby/setup-ruby/actions/runs/13167306524/job/36750485255#step:4:7) `macos-13` which changed that number, hence the error.

It's pretty annoying that I first need to guess that actual number since running on a specific macOS platform doesn't guarantee the compiler was built on that version but also that suddenly it could change and now I have to add that new platform in all my lockfiles whereas other platforms seems pretty deterministic for that value.

This is an issue for `x86_64-darwin` not only `arm64-darwin`.

**Comment by deivid-rodriguez** [comment]
@starburst997 Yeah, I can see how that can be annoying. I think we'll eventually deprecate `-NN` darwin platform versions in lockfile, and start ignoring the `-NN` part everywhere.

For now, we've taken some steps to alleviate this issue:
* Fresh lockfiles don't include `-NN` platforms by default, unless some locked gem has released a `-NN` variant matching your platform (but I'm not aware of any gems doing this).
* You can normalize your lockfile by running `bundle lock --normalize-platforms` or more specifically to only remove the inconvenient platform by running `bundle lock --remove-platform <platform-with-version>` and then `bundle lock --add-platform <platform-without-version>`).

**Comment by deivid-rodriguez** [comment]
The only explanation that I can think of is that you are using Ruby 3.5-dev, which is incompatible with the native gem, that's why I ask about your Ruby version.

**Comment by jeremywadsack** [comment]
Ruby version:

```
$ ruby --version
ruby 3.3.7 (2025-01-15 revision be31f993d7) [arm64-darwin23]
```

Gemfile:
```
# frozen_string_literal: true

ruby "3.3.7"

source "https://rubygems.org"

gem 'grpc'
```

Gemfile.lock:
```
GEM
  remote: https://rubygems.org/
  specs:
    bigdecimal (3.1.9)
    google-protobuf (4.30.0)
      bigdecimal
      rake (>= 13)
    google-protobuf (4.30.0-aarch64-linux)
      bigdecimal
      rake (>= 13)
    google-protobuf (4.30.0-arm64-darwin)
      bigdecimal
      rake (>= 13)
    google-protobuf (4.30.0-x86-linux)
      bigdecimal
      rake (>= 13)
    google-protobuf (4.30.0-x86_64-darwin)
      bigdecimal
      rake (>= 13)
    google-protobuf (4.30.0-x86_64-linux)
      bigdecimal
      rake (>= 13)
    googleapis-common-protos-types (1.18.0)
      google-protobuf (>= 3.18, < 5.a)
    grpc (1.70.1)
      google-protobuf (>= 3.25, < 5.0)
      googleapis-common-protos-types (~> 1.0)
    grpc (1.70.1-aarch64-linux)
      google-protobuf (>= 3.25, < 5.0)
      googleapis-common-protos-types (~> 1.0)
    grpc (1.70.1-arm64-darwin)
      google-protobuf (>= 3.25, < 5.0)
      googleapis-common-protos-types (~> 1.0)
    grpc (1.70.1-x86-linux)
      google-protobuf (>= 3.25, < 5.0)
      googleapis-common-protos-types (~> 1.0)
    grpc (1.70.1-x86_64-darwin)
      google-protobuf (>= 3.25, < 5.0)
      googleapis-common-protos-types (~> 1.0)
    grpc (1.70.1-x86_64-linux)
      google-protobuf (>= 3.25, < 5.0)
      googleapis-common-protos-types (~> 1.0)
    rake (13.2.1)

PLATFORMS
  aarch64-linux
  arm64-darwin
  ruby
  x86-linux
  x86_64-darwin
  x86_64-linux

DEPENDENCIES
  grpc

RUBY VERSION
   ruby 3.3.7p123

BUNDLED WITH
   2.6.5
```

But it sounds like bundler should be falling back from `arm64-darwin-23` to `arm64-darwin` and find that version of the gem anyway. So this might be a problem with something on my system, unrelated to this issue?

**Comment by deivid-rodriguez** [comment]
Another possible culprit, can you check `bundle config` output? If you have `force_ruby_platform` configured in there, that's another likely culprit, and removing that (`bundle config unset force_ruby_platform`) should do the trick.

**Comment by headius** [comment]
A similar issue exists for JRuby, which reports its `arch` as `universal-java-<version>`. See discussion in #3520.

I think the right fix across the board would be to ignore the platform version for such uses.

**Comment by hsbt** [comment]
`bundle lock --normalize-platforms` resolve the original report, I think.

**Comment by sandstrom** [comment]
@hsbt Great! 💯 

I can confirm it does:

<img width="430" height="310" alt="Image" src="https://github.com/user-attachments/assets/7505197d-959d-4f5a-b766-b034158ac0b2" />

**Comment by headius** [comment]
Were there any actual changes for this, and does it resolve my other issue with extensions being installed in version-specific dirs? #3520

**Comment by sandstrom** [comment]
If you search in this repo for the command flag referenced two comments above, maybe you'll find something. 

But the top comment that I reported (this issue) is solved, I'd say.

If you think something else is not working, maybe better to create another issue explaining it.

**Comment by headius** [comment]
It appears this issue only fixed the problem in `Gemfile.lock`, not on the filesystem where RubyGems indicates if extensions have been built. I have been forced to hack around the problem: https://github.com/ruby/rubygems/issues/3520#issuecomment-3519172989
